pkg/didcomm/protocol/route/service.go

```diff
@@ -296,6 +328,80 @@ func (s *Service) handleForward(msg *service.DIDCommMsg) error {
 	return s.outbound.Forward(forward.Msg, dest)
 }
 
+// Register registers the agent with a router with the on connectionID passed in. The agent is
+// registered with the router and retrieves the router endpoint and routing keys. This function
+// throws an error if the agent is already registered against a router. This function internally
+// works as a sync function and works in-conjunction with handleGrant function. ie, The request
+// did-comm message is sent to the router and it waits for the grant response from the router
+// till timeout.
+func (s *Service) Register(connectionID string) error {
+	// check if router is already registered
+	routerConnID, err := s.getRouterConnectionID()
+	if err != nil && !errors.Is(err, storage.ErrDataNotFound) {
+		return fmt.Errorf("fetch router connection id : %w", err)
+	}
+
+	if routerConnID != "" {
+		return errors.New("router is already registered")
+	}
+
+	// get the connection record for the ID to fetch DID information
+	conn, err := s.connectionLookup.GetConnectionRecord(connectionID)
+	if err != nil {
+		if errors.Is(err, storage.ErrDataNotFound) {
+			return ErrConnectionNotFound
+		}
+
+		return fmt.Errorf("fetch connection record from store : %w", err)
+	}
+
+	// generate message ID
+	msgID := uuid.New().String()
+
+	// register chan for callback processing
+	grantCh := make(chan Grant)
+
+	s.routeRegistrationMapLock.Lock()
+	s.routeRegistrationMap[msgID] = grantCh
+	s.routeRegistrationMapLock.Unlock()
```

Comment:
Move set/get to different function, just to avoid accidents in future

Reply:
[done](https://github.com/hyperledger/aries-framework-go/compare/4d6cf6d2117e03d7fa5fe3e7f5623c5d317dc7df..f2c71755a81cac382569f3cb14fae87828d66b2c#diff-7cc3df019152fe2c557826a312cc0a12R361).